fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexScrubbingToolsMissing.java

```diff
@@ -137,59 +125,78 @@ public CompletableFuture<Issue> handleOneItem(final FDBRecordStore store, final
                 });
     }
 
-    public CompletableFuture<Pair<MissingIndexReason, Tuple>> detectMissingIndexKeys(FDBStoredRecord<Message> rec) {
-        // return the first missing (if any).
+    @SuppressWarnings("PMD.CloseResource")
+    private CompletableFuture<Pair<MissingIndexReason, Tuple>> detectMissingIndexKeys(final FDBRecordStore store, FDBStoredRecord<Message> rec) {
+        // Generate synthetic record (if applicable) and return the first detected missing (if any).
+        final AtomicReference<Pair<MissingIndexReason, Tuple>> issue = new AtomicReference<>();
+
+        if (!isSynthetic) {
+            return checkMissingIndexKey(rec, issue).thenApply(ignore -> issue.get());
+        }
+        final RecordQueryPlanner queryPlanner =
+                new RecordQueryPlanner(store.getRecordMetaData(), store.getRecordStoreState().withWriteOnlyIndexes(Collections.singletonList(index.getName())));
+        final SyntheticRecordPlanner syntheticPlanner = new SyntheticRecordPlanner(store, queryPlanner);
+        SyntheticRecordFromStoredRecordPlan syntheticPlan = syntheticPlanner.forIndex(index);
+        final RecordCursor<FDBSyntheticRecord> recordCursor = syntheticPlan.execute(store, rec);
+
+        return AsyncUtil.whenAll(
+                recordCursor.asStream().map(syntheticRecord -> checkMissingIndexKey(syntheticRecord, issue))
+                        .collect(Collectors.toList()))
+                .whenComplete((ret, e) -> recordCursor.close())
+                .thenApply(ignore -> issue.get());
+
+    }
+
+    private CompletableFuture<Void> checkMissingIndexKey(FDBIndexableRecord<Message> rec,
+                                                         AtomicReference<Pair<MissingIndexReason, Tuple>> issue) {
+        // Iterate grouping keys (if any) and detect missing index entry (if any)
         final KeyExpression root = index.getRootExpression();
         final Map<Tuple, List<LuceneDocumentFromRecord.DocumentField>> recordFields = LuceneDocumentFromRecord.getRecordFields(root, rec);
         if (recordFields.isEmpty()) {
-            // Could recordFields be an empty map?
-            return CompletableFuture.completedFuture(Pair.of(MissingIndexReason.EMPTY_RECORDS_FIELDS, null));
+            // recordFields should not be an empty map
+            issue.compareAndSet(null, Pair.of(MissingIndexReason.EMPTY_RECORDS_FIELDS, null));
+            return AsyncUtil.DONE;
         }
         if (recordFields.size() == 1) {
-            // A single grouping key
-            return checkMissingIndexKey(rec, recordFields.keySet().stream().findFirst().get());
+            // A single grouping key, simple check.
+            return checkMissingIndexKey(rec, recordFields.keySet().stream().iterator().next(), issue);
```

Comment:
I don't think there's any reason to have `stream` here. But I'm not sure that there's much value to having a separate code path when there is only one grouping key. It avoids creating a future, but complicates the code. 
```suggestion
            return checkMissingIndexKey(rec, recordFields.keySet().iterator().next(), issue);
```

Reply:
✅